### DISCUSSION TITLE
Support Yarn 2 PnP

### DISCUSSION THREAD

**Discussion by beeequeue** [vercel/vercel#4223] [Ideas]
[Yarn Berry](https://github.com/yarnpkg/berry) is currently in release candidate status, and it's a good idea to start thinking about supporting it, as can bring some complications due to Plug N' Play being the default.

The current issue for me is that the `yarn install` command has had breaking changes, and in a repo with the dependencies committed it should run `yarn install --immutable --immutable-cache`, and `--check-cache` if you want to validate the dependencies before deploying. [Source](https://yarnpkg.com/cli/install)

There will probably have to be some changes to how the scripts are run as well to make sure that they support plug n play. [Read more](https://yarnpkg.com/advanced/migration#call-your-scripts-through-yarn-node-rather-than-node)

Other migration steps can be found [here](https://yarnpkg.com/advanced/migration)

----

For now, would it be possible to pass a custom install script in the configuration? It has been suggested in other discussions as well.

**Comment by styfle** [comment]
Thanks for the suggestion!

We might add an option for "Install Command" in [Project Settings](https://zeit.co/blog/advanced-project-settings), however we would need a way to indicate that yarn 2 should be installed before the install script is run, or perhaps it could be apart of the script itself.

This is similar to the suggestion in #3896 which has a workaround you can use by modifying your [Build Command](https://github.com/zeit/now/discussions/3896#discussioncomment-1874).

**Reply by beeequeue** [reply]
Yarn 2 doesn't have to be installed in the current recommended setup - all yarn 2 files including yarn 2 are committed in the `.yarn` folder. You only need a recent yarn 1.x version installed.

Dependencies are also committed in here (`.yarn/cache`), so you don't actually have to run `yarn install` to run your application, but you can do it to check that all the dependencies haven't been messed with (`yarn install --check-cache`).

As far as I could tell the workaround sadly doesn't work as you never get to the `postinstall` when it fails on the `install` script.

You can check out my setup as an example: https://github.com/BeeeQueue/cr-proxy/tree/yarn-2

In should in theory be runnable after just checking out the branch.

**Comment by jakst** [comment]
Any updates on this? It would be very nice if Vercel supported the latest ecosystem tooling

**Reply by kachkaev** [reply]
Some relevant links I’ve seen so far:

- https://vercel.com/knowledge/does-vercel-support-yarn-2
- https://github.com/vercel/vercel/issues/5280
- https://github.com/vercel/vercel/discussions/4910

My own attempt to upgrade an Vercel/Next app to Yarn v2: https://github.com/kachkaev/njt/pull/29

**Comment by mormahr** [comment]
I just tried to migrate to yarn v2 and while installation step works, the vercel deployment service doesn't pick up that `next` is installed, because there's no `node_modules`. 

```Error: No Next.js version could be detected in your project. Make sure `"next"` is installed in "dependencies" or "devDependencies"```

Locally on my machine I tried `yarn dev` and `yarn build` and both worked.
It should be noted that yarn v2 doesn't use `node_modules` anymore, so everything has to be started through yarn (because it configures node as neccesary) or be explicitly use the [PnP API](https://yarnpkg.com/advanced/pnpapi). While using yarn to start the frontend build shouldn't be a problem, i think the serverless routes need a little more work to be compatible. Iirc for those routes vercel uses `ncc`, so that tool might need to be made aware of [pnp](https://yarnpkg.com/features/pnp).

**Reply by armandabric** [reply]
There is another discussion about issue with Yarn 2 where the issue is detailed: https://github.com/vercel/vercel/discussions/4910

**Comment by TooTallNate** [comment]
Hello friends. I wanted to notify you that we have added support for Yarn v2 in our official Runtimes (`@vercel/next`, `@vercel/node` and `@vercel/static-build`, specifically, see https://github.com/vercel/vercel/pull/5552). Currently the update is available on the "canary" release channel.

If you would like to try out the Yarn v2 support before it hits the stable release channel, you can run the following command if you are using the CLI:

```
$ vc -b FORCE_BUILDER_TAG=canary
```

If you are using the Git integration to trigger your deployments, then you can define the `FORCE_BUILDER_TAG` environment variable in your Project's environment variables setting on the Vercel dashboard.

Please let me know if it works for you or if you bump into any issues!

**Reply by osdiab** [reply]
As commented in the linked PR (https://github.com/vercel/vercel/pull/5552#issuecomment-748602845 ), seems like it’s not really yarn Berry being supported, but rather forcefully bypassing PnP.  Is proper yarn v2 support in the Vercel pipeline? Thanks for the great platform nonetheless!

**Reply by jensmeindertsma** [reply]
@TooTallNate when will this feature become available on stable and the Git integration without any configuration?

**Comment by barbogast** [comment]
We are using yarn2 on Vercel without any issues with `getServerSideProps` and `getStaticProps` as well as `revalidate: 1`. Didn't do any particular setup, didn't notice any issues.

_Edit_: We are using PnP, seems to work perfectly fine.

**Comment by kidonng** [comment]
I think Vercel should make it clear that **it works with Yarn 2+, as long as you don't use [PnP](https://yarnpkg.com/features/pnp/)**.

And Vercel should also cache `.yarn/cache` if it exists.

**Reply by barbogast** [reply]
PnP works fine for us.

**Reply by kidonng** [reply]
Are you referring to static builds (e.g. Next.js)? That works because Vercel forces Yarn 2+ to disable PnP (see #5552), and you are likely to see linking step in the logs, which PnP doesn't need.
As for Serverless Functions with PnP, it completely breaks, at least that's how it went when I tried last time.

**Reply by barbogast** [reply]
Ou, right, I didn't notice that part. Yeah, there is a link step. This explains my confusion about the vercel docs saying that yarn2 is not supported.

Still, for us it works, which is already a big deal. Sure, there is a risk of stuff breaking on vercel that works locally, but it's already a good step.

**Reply by belgattitude** [reply]
@kidonng 

Yes it seems it even works with yarn 3.0.0-rc2 without pnp (`nodeLinker: 'node-modules'`). 

I'm playing a bit with monorepo setup https://github.com/belgattitude/nextjs-monorepo-example/pull/36 and haven't had any issues.

Except:

> And Vercel should also cache .yarn/cache if it exists.

Yes that would be awesome :)

**Reply by kidonng** [reply]
> > And Vercel should also cache .yarn/cache if it exists.
> 
> Yes that would be awesome :)

Maybe we should just open a new issue about this😄

**Reply by kidonng** [reply]
@belgattitude I think I found a trick here: Vercel caches the entire `node_modules`, so for it to cache Yarn 2/pnpm/whatever, just put cache there.

Will try experiment a bit if this works for Yarn 2:

```yaml
cacheFolder: "./node_modules/.yarn-cache"
```

**Reply by beeequeue** [reply]
If this works it would be better to set `YARN_CACHE_FOLDER=./node_modules/.yarn-cache` in the Vercel environment variables, so it only happens when when building it on the server and not on your computer

**Reply by belgattitude** [reply]
@kidonng 

> Maybe we should just open a new issue about thissmile

Did yesterday, but could not find how to create a github issue, was redirected to support.

> cacheFolder: "./node_modules/.yarn-cache"

Possibly yes, mmm node_modules/.xxx... there might be edge cases now or in the future... Let me know anyway, I might try it out too. But it would double the cache size somehow :(

Thanks

**Reply by belgattitude** [reply]
@BeeeQueue 

> If this works it would be better to set YARN_CACHE_FOLDER=./node_modules/.yarn-cache in the Vercel environment

Also possible to do it from the command line a bit like they did for pnpm https://github.com/vercel/next.js/issues/16471#issuecomment-841391333

**Reply by belgattitude** [reply]
@BeeeQueue @kidonng Thanks for your tips... I've ended up overriding install from vercel setings:

```
YARN_CACHE_FOLDER=./node_modules/.yarn-cache yarn install
```

<img width="250" src="https://user-images.githubusercontent.com/259798/118364486-e2396200-b598-11eb-9f7b-1706fd9cfea5.png" />




PS/ prefer using env `YARN_CACHE_FOLDER` rather than `yarn install --cache-folder` as the later has been removed in yarn 3.0.0-rc.2

**Reply by belgattitude** [reply]
Edit: it seems to work now

~But actually not :smile: , still some edge cases: https://github.com/yarnpkg/berry/issues/2900 and https://github.com/belgattitude/nextjs-monorepo-example/pull/48.~

**Reply by belgattitude** [reply]
Caching in .next/cache seems to actually work and is not creating issues with yarn 3 (https://github.com/yarnpkg/berry/issues/2900#issuecomment-860049252)

So you can override install by running `YARN_CACHE_FOLDER=./.next/cache/yarn-cache yarn install`.

![image](https://user-images.githubusercontent.com/259798/121776627-2f0b5b00-cb8e-11eb-9090-19f5cafc6522.png)

And the magic happen: 

![image](https://user-images.githubusercontent.com/259798/121776561-db007680-cb8d-11eb-94cc-ea3fd3e50853.png)

Tested on a monorepo but should work for regular apps as well.

**Comment by belgattitude** [comment]
Hey Adam, Kid

the YARN_CACHE_FOLDER=./node_modules/.yarn-cache is nice and works with
pnpm, but actually does not work when no (or not yet) cache is present...

I've opened an issue https://github.com/yarnpkg/berry/issues/2900 and a
repro https://github.com/belgattitude/nextjs-monorepo-example/pull/48.

My feeling about this: vercel should be able to detect yarn 2 and
nodeLinker node_modules from .yarnrc.yml or metadadata keys in yarn.lock...
A call to `yarn config get cacheFolder` would help in that case and could
be used to upload the cache content. But I understand it would add more
complexity to vercel.

In the meantime, I'll try to set up a script to set YARN_CACHE_FOLDER only
if a `nodes_modules/.yarn-state.yml` exists and hope yarn will allow this
in the future.

Thanks you for replies

Happy week

Sébastien

**Comment by mareksuscak** [comment]
For those of you who're still struggling with Yarn v4, which changes the global cache behavior compared to v3, you have two options:

1. Override the "Install Command" in Settings -> General and use the following command:
  
    ```
    YARN_CACHE_FOLDER=node_modules/.yarn-cache YARN_ENABLE_GLOBAL_CACHE=false yarn install
    ```

2. Add the following 2 env vars in Settings -> Environment Variables:

    ```
    YARN_CACHE_FOLDER=node_modules/.yarn-cache
    YARN_ENABLE_GLOBAL_CACHE=false
    ```

    Tip: Just copy and paste the above vars, Vercel will automatically create two entries ;)

The longer explanation is that Yarn v4 [forces the use of the global cache folder](https://yarnpkg.com/configuration/yarnrc#enableGlobalCache) over the configured cache folder which means that it won't work correctly unless you disable the global cache in addition to setting the cache folder explicitly which is somewhat unintuitive.
